scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py

```diff
@@ -0,0 +1,131 @@
+from time import sleep
+from typing import Dict, List
+
+class Maker:
+    '''
+    The Maker class represents the coffee maker itself. It keeps track of the water, milk,
+    and beans in the machine. Makes coffee and reports on the resources.
+    '''
+
+    def __init__(self, resources: Dict [str, int]):
+        self.resources = resources
+
+    def report(self):
+        for resource, amount in self.resources.items():
+            print(f'{resource}: {amount} ml')
+
+    def is_resource_sufficient(self, menu_item):
+        for ingredient, amount in menu_item.ingredients.items():
+            if amount >= self.resources.get(ingredient, 0):
+                print(f'Sorry there is not enough {ingredient}.')
+                print()
+                return False
+        return True
+
+    def make_coffee(self, menu_item):
+        for ingredient, amount in menu_item.ingredients.items():
+            self.resources[ingredient] -= amount
+            sleep(5)
+        print(f"Here's your {menu_item.name}. Enjoy! :)")
+        print()
+
+
+class MenuItem:
+    '''
+    The MenuItem class represents a menu item. It keeps track of the name, the ingredients,
+    and the cost.
+    '''
+
+    def __init__(self, name: str, cost: float, ingredients: Dict[str, int]):
+        self.name = name
+        self.cost = cost
+        self.ingredients = ingredients
+
+    def __str__(self):
+        return f'{self.name.capitalize()} (${self.cost})'
+
+
+class Menu:
+    '''
+    The Menu class represents the menu. It keeps track of the menu items where each item is
+    an instance of the MenuItem class. The class holds the functionalities for finding an
+    individual menu item and getting all the menu items.
+    '''
+
+    def __init__(self, menu_items: List[MenuItem]):
+        self.menu_items = menu_items
+
+    def get_all_menu_items(self):
+        options = ''
+        for item in self.menu_items:
+            options += f'{item.name}/'
+        return options
+
+    def find_menu_item(self, item_name: str):
+        for item in self.menu_items:
+            if item.name == item_name:
+                return item
+        print('Sorry that item is not available.')
+
+class Stash:
+    '''
+    The Stash class represents the stash. It keeps track of the money in the machine.
+    '''
+    def __init__(self, profit: int):
+        self.profit = profit
+
+
+    def report(self):
+        print(f'Money: ${self.profit}')
+
+    def make_payment(self, price: float):
+        money_received = float(input('How much money are you giving me? '))
```

Comment:
This needs `try/except` to ensure number is entered.

Reply:
I have added exception handling here as well as in the `find_menu_item` method. Previously in cases where the method returns `None` the print statement in the main section used to fail. Now the method throws an exception. I thought about doing something like creating a custom exception like "ItemNotFoundError" but went with a plain exception at the end. Please let me know your thoughts on that too.